#Домашнє завдання до тем apply(), groupby(), pivot_table()

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [1]:
import pandas as pd
df = pd.read_csv('../../data/supermarket_sales.csv')

1. Дослідимо, який філіал супермаркету ('Branch') є найприбутковішим. Для цього знайдіть сумарний прибуток за кожним філіалом і виявіть, який філіал має найвищий.

In [4]:
df.groupby('Branch')['gross income'].sum()

Branch
A    5057.1605
B    5057.0320
C    5265.1765
Name: gross income, dtype: float64

`С` – є найприбутковішим філіалом супермаркету.

2. В якому місті знайходиться філіал з найвищим прибутком? Може в тому місці нам розмітисти ще один магазин.  
Знайдіть відповідь за допомогою функціоналу Pandas.

In [39]:
# 1й спосіб
df_sorted = df.sort_values('gross income', ascending = False).head(1)

df_sorted

Invoice ID Branch       City Customer type  Gender         Product line  \
350  860-79-0874      C  Naypyitaw        Member  Female  Fashion accessories   

     Unit price  Quantity  Tax 5%    Total       Date   Time      Payment  \
350        99.3        10   49.65  1042.65  2/15/2019  14:53  Credit card   

      cogs  gross margin percentage  gross income  Rating  
350  993.0                 4.761905         49.65     6.6

In [41]:
df_sorted.groupby('City')[['Branch', 'gross income']].max()

Branch  gross income
City                          
Naypyitaw      C         49.65

In [45]:
# 2й спосіб
branch_income = df.groupby('Branch')['gross income'].sum()

top_branch = branch_income.idxmax()

In [47]:
top_branch

'C'

In [49]:
top_city = df[df['Branch'] == top_branch]['City'].iloc[1]
top_city

'Naypyitaw'

In [51]:
df.groupby('City').size()

City
Mandalay     332
Naypyitaw    328
Yangon       340
dtype: int64

У місті `Naypyitaw` знаходиться філіал з найвищим прибутком - 5265.1765. Оскільки у цьому місті 328 магазинів, тому можна ще розмістити декілька, адже в інших містах їх більше (332 і 340 відповідно).

3.1. Створіть зводну таблицю, яка покаже, скільки покупок (інвойсів) було зроблено в кожній з філій (`Branch`) за різними категоріями товарів. Запишіть таблицю в змінну `invoices_by_category` і виведіть змінну на екран.
Ця таблиця допоможе проаналізувати, в якій філії купують найбільше товарів кожної з категорій.

In [26]:
invoices_by_category = pd.pivot_table(df, index='Branch', columns='Product line', values='Invoice ID', aggfunc='size')

In [28]:
invoices_by_category

Product line  Electronic accessories  Fashion accessories  Food and beverages  \
Branch                                                                          
A                                 60                   51                  58   
B                                 55                   62                  50   
C                                 55                   65                  66   

Product line  Health and beauty  Home and lifestyle  Sports and travel  
Branch                                                                  
A                            47                  65                 59  
B                            53                  50                 62  
C                            52                  45                 45

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1rueAdko6S3UxIHGtojetTxlES-EyM6Yb)

3.2. Викристовуючи змінну `invoices_by_category` дайте відповідь програмно (тобто значення треба не просто знайти очима, а вивести за допомогою коду), в якому філіалі магазину (`Branch`) найбільше інвойсів із покупкою товарів категорії "Електронні аксесуари" (`Electronic accessories`)?


In [112]:
invoices_by_category.idxmax()[:1]

Product line
Electronic accessories    A
dtype: object

4-6. **Творче завдання на розвиток аналітичного мислення**

Крок 1. Сформулюйте ТРИ питання (гіпотези) до наявних даних, які допомогли б вам зрозуміти, які користувачі що, де та коли найбільше/найменше купують, аби дати на основі цих гіпотез рекомендації бізнесу. Звісно питання мають бути не тими, на які ми вже відповіли в завданнях модулю.

Крок 2. Знайдіть відповіді на свої питання з допомогою функціоналу pandas.

Крок 3. Напишіть, як відповідь на це питання може бути використана для прийняття бізнес рішень.   
   
 Питання можуть бути будь-якої складності, але їх має бути 3. Кожне питання оцінюється як 1 завдання. Без виконання цього завдання ДЗ не приймається. Якщо є питання щодо виконання - пишіть у чат 🙌


**1. Яка категорія покупців за статтю (чоловіки/жінки) купує більше товарів у кожній категорії?**

**Гіпотеза:** Чоловіки частіше купують товари певних категорій (наприклад, електроніка), тоді як жінки — інші (наприклад, одяг, косметику чи продукти для дому).

**2. Чи впливає наявність картки лояльності на середній чек покупця?**

**Гіпотеза:** Покупці з карткою лояльності витрачають більше за одну покупку, користуючись знижками чи бонусами, ніж ті, хто її не має.

**3. Чи впливає тип оплати на суму чеку?**

**Гіпотеза:** Користувачі, які оплачують карткою, здійснюють покупки на більші суми, ніж ті, хто платить готівкою або електронним гаманцем.

In [118]:
df.columns

Index(['Invoice ID', 'Branch', 'City', 'Customer type', 'Gender',
       'Product line', 'Unit price', 'Quantity', 'Tax 5%', 'Total', 'Date',
       'Time', 'Payment', 'cogs', 'gross margin percentage', 'gross income',
       'Rating'],
      dtype='object')

In [153]:
#Гіпотеза 1

df.groupby(['Gender', 'Product line'])['gross income'].sum()

Gender  Product line          
Female  Electronic accessories    1290.5725
        Fashion accessories       1449.4000
        Food and beverages        1579.5675
        Health and beauty          883.8565
        Home and lifestyle        1430.3275
        Sports and travel         1360.7010
Male    Electronic accessories    1296.9290
        Fashion accessories       1136.5950
        Food and beverages        1093.9965
        Health and beauty         1458.7025
        Home and lifestyle        1134.5255
        Sports and travel         1264.1955
Name: gross income, dtype: float64

In [243]:
filtered_electronic = df[df['Product line']=='Electronic accessories']
filtered_electronic.Gender.value_counts(normalize=True)

Gender
Male      0.505882
Female    0.494118
Name: proportion, dtype: float64

In [245]:
filtered_food = df[df['Product line']=='Food and beverages']
filtered_food.Gender.value_counts(normalize=True)

Gender
Female    0.517241
Male      0.482759
Name: proportion, dtype: float64

In [247]:
filtered_health_beauty = df[df['Product line']=='Health and beauty']
filtered_health_beauty.Gender.value_counts(normalize=True)

Gender
Male      0.578947
Female    0.421053
Name: proportion, dtype: float64

1. `Електронні аксесуари` купують майже однаково як жінки, так і чоловіки - 1290.57 та 1296.92 відповідно (49% і 50%).

**Рекомендації:** Розширити асортимент, який буде однаково цікавий обом статям (наприклад, стильні та функціональні аксесуари для жінок і практичні гаджети для чоловіків).
Використати універсальний маркетинг, орієнтований на спільні інтереси.

2. `Їжа та напої` - жінки частіше купують цю категорію (51% проти 48%).

**Рекомендації:** Розширення асортименту, який приверне увагу саме чоловіків (продукти для спорту, снеки для активного способу життя або делікатеси, знижки на пивні набори тощо).

3. `Здоровʼя та краса`. Чоловіки значно більше купують цю категорію (57% проти 42%).

**Рекомендації:** Розширити асортимент товарів, які цікавлять жінок: косметика - натуральні та органічні засоби догляду, вітаміни та харчові добавки, орієнтовані на жіноче здоровʼя, догляд за волоссям, шкірою та тілом. Додати вигідні пропозиції, бонуси, кешбек чи знижки на популярні товари.

In [206]:
#Гіпотеза 2

df.groupby('Customer type')['Total'].mean()

Customer type
Member    327.791305
Normal    318.122856
Name: Total, dtype: float64

Як бачимо, середнє значення чеку для тих клієнтів, хто має карту лояльності вище - 327.79, ніж для звичайних покупців - 318.12.
**Рекомендація:** стимулювати клієнтів отримати картку лояльності через вигідні пропозиції (додаткові знижки, акції, кешбек).

In [473]:
#Гіпотеза 3

top_payment = df.groupby(['Payment', 'City'])['Total'].max()

In [475]:
top_payment.idxmax()

('Credit card', 'Naypyitaw')

In [477]:
top_payment

Payment      City     
Cash         Mandalay      867.090
             Naypyitaw    1003.590
             Yangon        932.337
Credit card  Mandalay     1022.385
             Naypyitaw    1042.650
             Yangon       1039.290
Ewallet      Mandalay     1022.490
             Naypyitaw    1034.460
             Yangon        951.825
Name: Total, dtype: float64

In [479]:
df.Payment.value_counts(normalize=True)

Payment
Ewallet        0.345
Cash           0.344
Credit card    0.311
Name: proportion, dtype: float64

Вищі суми чеків оплачують карткою у місті `Naypyitaw`. Але загалом по всім містам та покупкам розподіл між готівкою та електронним гаманцем майже рівномірний по 34%, а кредитна картка - 31%. Тож є потенційно цікавість до нових технологій.

**Рекомендація:** Клієнти, які платять карткою, можуть бути більш схильними до великих покупок. Можна стимулювати оплату карткою, пропонуючи знижки або бонуси для карткових покупців. 
Варто також розглянути можливість впровадження додаткових варіантів оплати через електронні гаманці.